### ISSUE TITLE
DefaultDomRenderer2 overrides custom element's class

### ISSUE THREAD

**Issue by hotforfeature** [angular/angular#15399]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Using a custom element that adds a class to itself is incompatible when setting the class in a component template.

**Expected behavior**
Classes set in a component template should be added to the classList, instead of overriding it.

**Minimal reproduction of the problem with instructions**
- Create a custom element that modifies its classList in `connectedCallback()`
- Use the custom element with a class attribute in a component template

View the demo in Chrome (native webcomponent support)
http://plnkr.co/edit/4pzXr214xEZyUZ9ftLG7?p=preview

**What is the motivation / use case for changing the behavior?**
Feeling the love between Angular and Polymer 2.

Specifically, the `<paper-button>` element adds the `.paper-material` class to its classList [when attached to the DOM](https://github.com/PolymerElements/paper-button/blob/2.0-preview/paper-button.html#L170).

Angular first [creates and attaches the element](https://github.com/angular/angular/blob/master/packages/core/src/view/element.ts#L147) (and `<paper-button>` adds its class hooray!).

Next, the function asks the renderer [to set the attributes from the component markup](https://github.com/angular/angular/blob/master/packages/core/src/view/element.ts#L167).

In the browser, the [renderer uses `setAttribute()`](https://github.com/angular/angular/blob/master/packages/platform-browser/src/dom/dom_renderer.ts#L151) on the element. When calling `element.setAttribute('class', 'name');`, existing classes are removed.

**Suggested Fix**

I think it makes the most sense for `DefaultDomRenderer2` to understand that setting the class attribute can have negative consequences.

```ts
export class DefaultDomRenderer2 implements Renderer2 {
 ...
 setAttribute(el: any, name: string, value: string, namespace?: string): void {
    if (namespace) {
      el.setAttributeNS(NAMESPACE_URIS[namespace], namespace + ':' + name, value);
    } else if (name === 'class') {
      value.split(' ').forEach(className => {
        el.classList.add(name, className);
      });
    } else {
      el.setAttribute(name, value);
    }
  }
}
```

* **Angular version:** 2.4.10+

* **Browser:** [all]
 
* **Language:** [TypeScript 2.0.2]

* **Node (for AoT issues):** `node --version` =   v6.9.5

**Comment by DzmitryShylovich** [comment]
Duplicate of https://github.com/angular/angular/issues/7289

**Comment by hotforfeature** [comment]
I disagree. While the concepts are the same (conflicting class assignments), that issue deals with Angular components. This issue is about Webcomponents, which are completely separate from Angular. 

I'd be interested to see if my suggestion resolves any of the problems in #7289 though. I'll have to test that out.

**Comment by pkozlowski-opensource** [comment]
A work-around for now: `<custom-element [class.override]="true"></custom-element>`: http://plnkr.co/edit/yJTXxOAJeqnQ1RXbhmg2?p=preview

**Comment by robdodson** [comment]
@domenic I'm curious what your take on this is. Should a Custom Element be allowed to set its own classes or should it do everything through attributes?

**Comment by domenic** [comment]
I think class is explicitly meant to be used by consumers, and a custom element should not modify its own classes.

**Comment by robdodson** [comment]
I spoke with the Polymer team about this today. They said they agreed that an element setting its own class was too easy to get stomped on and they no longer do it. While they used to set `.paper-material` in an earlier version of the 2.0 preview, they've switched off of that. I think the one exception is a `.keyboard-focus` class left over from an a11y input modality polyfill.

**Comment by pkozlowski-opensource** [comment]
tl;dr; this is fixed / works as expected in Angular v9

As of Angular v9 (running ivy rendering engine, which is the default) CSS classes on a host element will be merged in this particular scenario: https://ng-run.com/edit/Xo9Rx1HX6ZKvwdV8cGGQ?open=app%2Fapp.module.ts

This comes down to the fact that attributes (here `class="override"`) are set on an element by Angular _before_ a given element is added to the DOM (so before the `connectedCallback()` fires).

I believe that the Angular v9 behaviour is more in-line with people expectations and solves this particular scenario (plus some others, ex.: #13699 and #18909). 

Still, as other commented, when we start to set attributes / properties on a host element (which is shared among the custom element implementation _and_ can be manipulated by Angular / other frameworks) things might get hairy. But this is the whole different conversation.
